### PR TITLE
Convolutional Layer in CUDA

### PR DESCRIPTION
**Pull request by steremma**
This is a PR including a lot of work on the convolutional layer.

It can be logically divided the following parts:

1. Back propagation in CUDA (this was the most complex)

2. Forward propagation in CUDA (this was a comparatively simpler task)

3. Helper static methods such as a column based sum reduction and matrix equality checkers for every architecture.

4. An element by element detailed test case for backward and forward propagation. The one on forward propagation was designed and solved on paper. I was not able to do the same for back-propagation, instead I considered the CPU implementation to be correct and asserted that the CUDA outputs the same result.

5. Trivial bug fixes.

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsft-bot  build

File: tmva/tmva/inc/TMVA/DNN/Architectures/Cuda/CudaMatrix.h [hunk 1]
```diff
@@ -151,6 +151,7 @@ class TCudaMatrix
    size_t GetNrows() const {return fNRows;}
    size_t GetNcols() const {return fNCols;}
    size_t GetNoElements() const {return fNRows * fNCols;}
+   size_t GetNElements() const {return fNRows * fNCols;}
```
Comment (lmoneta) [inline tmva/tmva/inc/TMVA/DNN/Architectures/Cuda/CudaMatrix.h]: Actually It would be better to keep GetNoElements in CudaMatrix and change in CpuMatrix. This however could be another PR

Comment (steremma) [reply tmva/tmva/inc/TMVA/DNN/Architectures/Cuda/CudaMatrix.h]: Indeed I remember our discussion, I just didn't want to further complicate the present PR by changing such a global module (`GetNelements()` is used in a lot of files, such as `MethodDL`).

I will submit the fix soon after merging this one
